File: tests/e2e/sre_authority/migration.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

import json

from .common import (
    AGENT, EPOCH, OPERATORS, PRIVATE, RUNTIME, SYSTEM, TENANT,
    assert_claim, assert_denial, enrollment_json, fingerprint, printed_object, require, review_args,
)
from .fixtures import CONTROL, CONTROL_NS, GROUP_BINDING, delegate_operators
from .admission import admission_cases, policies_ready
from .credential_paths import block_prestaged_paths, fixture_review, secret_watch, token_secret_denials


def cli_failure(result, text, label):
    require(result.returncode != 0 and text in (result.stdout + result.stderr),
            f"{label}: command did not fail for the intended reason")


def pods(h, namespace):
    return json.loads(h.k("get", "pods", "-n", namespace, "-o", "json"))["items"]


def snapshot_grants(h, spec):
    return {binding["name"]: h.get(binding["kind"].lower(), binding["name"], binding.get("namespace"))
            for binding in spec["legacyBindings"]}


def legacy_migration(h):
    policies_ready(h)
    delegate_operators(h)
    # Group grants were deliberately created before admission, never smuggled
    # through the new deny policies.
    original = h.get("clusterrolebinding", "kars-sre-reader")
    cli_failure(h.cli("authority", "preview", user="registrar", expected=None),
                "broad group grant", "broad-group review")
    require(h.get("clusterrolebinding", "kars-sre-reader")["subjects"] == original["subjects"],
            "Group ambiguity mutated direct legacy subjects")
    require(h.get("secret", PRIVATE, RUNTIME) is None, "Private material appeared before explicit enrollment")
    h.k("delete", "clusterrolebinding", GROUP_BINDING, "--wait=true", "--timeout=30s")
    h.passed("Actual preview blocks an unsafe pre-existing group grant before direct grants or credentials change")
    cli_failure(h.cli("authority", "preview", user="registrar", expected=None),
                "broad group grant", "watch-only broad-group review")
    assert_denial(h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets/router-services-admin",
                       user="watch-only"), "watch-only principal GET")
    assert_denial(h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets",
                       user="watch-only"), "watch-only principal LIST")
    secret_watch(h, CONTROL_NS, "e2e-watch-only-before", user="watch-only", allowed=True)
    h.passed("Watch-only group member is denied Secret GET/LIST but genuinely observes a newly created dummy Secret; preview blocks the grant")

    h.k("rollout", "status", f"deployment/{CONTROL}", "-n", CONTROL_NS, "--timeout=150s", timeout=160)
    spec = fixture_review(h)
    require(spec["sandbox"]["uid"] == h.state["legacy_source_uid"], "Staging adopted a replacement SRE source")
    require(spec["runtimeNamespace"]["uid"] == h.state["legacy_namespace_uid"], "Staging replaced the runtime namespace")
    assert_claim(h.get("karssandbox", "sre", SYSTEM), h.get("namespace", RUNTIME),
                 h.get("namespace", SYSTEM), h.state["system_uid"])
    require(spec.get("legacyConsumer") and len(spec["legacyBindings"]) >= 2, "Legacy reviews are incomplete")
    h.state["old_agent_sa_uid"] = h.get("serviceaccount", "sandbox", RUNTIME)["metadata"]["uid"]
    old_pods = {pod["metadata"]["uid"] for pod in pods(h, RUNTIME)}
    old_control_pods = {pod["metadata"]["uid"] for pod in pods(h, CONTROL_NS)}
    require(old_pods and old_control_pods, "Legacy/control consumers are not actually running")
    h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets/router-services-admin", user="old-agent", status=200)
    h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets", user="old-agent", status=200)
    before = snapshot_grants(h, spec)
    admission_cases(h, spec)
    token_secret_denials(h, before_enrollment=True)
    # The controller rejects the pre-existing token alias first, then the
    # watch-only group grant. Only explicit UID/RV-fenced fixture cleanup
    # removes them; the registration is left blocked on an incomplete review.
    blocked = block_prestaged_paths(h, spec, before)
    cli_failure(h.cli("authority", "enroll", "--sandbox-uid", spec["sandbox"]["uid"],
                     "--namespace-uid", spec["runtimeNamespace"]["uid"],
                     user="registrar", expected=None), "Every legacy binding", "missing binding reviews")
    current = h.get("karssreregistrations.kars.azure.com", "canonical")
    require(current["metadata"]["uid"] == blocked["metadata"]["uid"] and current["spec"] == blocked["spec"]
            and current["metadata"]["generation"] == blocked["metadata"]["generation"],
            "Unreviewed CLI enrollment changed the blocked registration")
    require(snapshot_grants(h, spec) == before, "Unreviewed enrollment mutated legacy bindings")
    require(h.get("deployment", "sre", RUNTIME)["spec"]["replicas"] == 1, "Unreviewed enrollment stopped the consumer")
    h.passed("Unreviewed bindings/consumer cannot enroll or mutate an occupied legacy source")
    require(snapshot_grants(h, spec) == before, "Controller mutated grants before validating the entire review set")
    require(h.get("deployment", "sre", RUNTIME)["spec"]["replicas"] == 1, "Controller stopped an unreviewed consumer")
    require(h.get("secret", PRIVATE, RUNTIME) is None, "Controller issued private material for an incomplete review")
    h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets/router-services-admin", user="old-agent", status=200)
    h.passed("Real registration/controller rejects an incomplete review before grant retirement, consumer stop or private issuance")

    # Refresh once after read-only tests; use the exact current RVs, not guessed
    # names or stale values from the fixture bootstrap.
    spec = enrollment_json(h.cli("authority", "preview", user="registrar"))
    reviewed = review_args(spec, current)
    preview = h.cli("authority", "enroll", *reviewed, "--dry-run", user="registrar")
    require(printed_object(preview)["spec"] == spec, "Enrollment dry-run differs from reviewed API identities")
    after = h.get("karssreregistrations.kars.azure.com", "canonical")
    require(after["spec"] == current["spec"] and after["metadata"]["generation"] == current["metadata"]["generation"]
            and after["metadata"]["uid"] == current["metadata"]["uid"], "Enrollment dry-run mutated registration")
    reviewed = review_args(spec, after)
    h.cli("authority", "enroll", *reviewed, user="registrar")
    require(h.get("karssreregistrations.kars.azure.com", "canonical")["spec"] == spec,
            "Persisted enrollment differs from the exact reviewed identities")

    issuance_seen = False
    def migrated():
        nonlocal issuance_seen
        private = h.get("secret", PRIVATE, RUNTIME)
        if private and "kube-token" in private.get("data", {}) and not issuance_seen:
            # Real held bearer token, with no client certificate fallback.
            assert_denial(h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets/router-services-admin",
                               user="old-agent"), "old held principal GET before private issuance")
            assert_denial(h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets",
                               user="old-agent"), "old held principal LIST before private issuance")
            assert_denial(h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets?watch=true&timeoutSeconds=1",
                               user="old-agent"), "old held principal WATCH before private issuance")
            require(not old_pods.intersection(pod["metadata"]["uid"] for pod in pods(h, RUNTIME)),
                    "Private credentials issued while a reviewed old consumer still exists")
            issuance_seen = True
        reg = h.get("karssreregistrations.kars.azure.com", "canonical")
        require(reg is not None, "Enrollment disappeared during migration")
        status = reg.get("status", {})
        require(not (status.get("phase") == "Blocked" and status.get("observedGeneration") == reg["metadata"]["generation"]),
                "Reviewed SRE migration was blocked")
        return reg if status.get("phase") == "Ready" and status.get("observedGeneration") == reg["metadata"]["generation"] else False
    reg = h.poll("reviewed migration and credential-order proof", migrated, seconds=240, interval=0.25)
    require(issuance_seen, "No private issuance was observed")
    h.cli("authority", "migrate", user="registrar", timeout=45)
    h.wait_ready()
    h.passed("Real old-principal GET/LIST/WATCH denial and old Pod termination precede private credential issuance")
    for namespace, cluster, name in [
        (CONTROL_NS, False, "e2e-old-watch-control"),
        (RUNTIME, False, "e2e-old-watch-private"),
        (CONTROL_NS, True, "e2e-old-watch-cluster"),
    ]:
        secret_watch(h, namespace, name, cluster=cluster)
    h.passed("Held legacy token receives actual HTTP Forbidden for namespaced and cluster Secret watches, including newly created dummy-Secret probes")
    require(h.get("serviceaccount", "sandbox", RUNTIME)["metadata"]["uid"] == h.state["old_agent_sa_uid"],
            "Migration used ServiceAccount recreation instead of revoking held-principal authorization")
    reader = h.get("clusterrolebinding", "kars-sre-reader")
    require(h.state["unrelated_subject"] in reader["subjects"], "Migration removed an unrelated binding subject")
    require(not any(subject.get("name") == "sandbox" and subject.get("namespace") == RUNTIME for subject in reader["subjects"]),
            "Legacy SRE subject remains bound")
    h.api("GET", f"/api/v1/namespaces/{CONTROL_NS}/secrets/router-services-admin", user="unrelated", status=200)
    h.passed("Unrelated binding subjects remain authorized; the old Sandbox ServiceAccount UID is preserved but denied")

    rotated = h.get("secret", "router-services-admin", CONTROL_NS)
    require(rotated["metadata"]["uid"] == h.state["control_secret_uid"], "Owned control Secret was replaced instead of rotated")
    require(fingerprint(rotated) != h.state["control_digest"], "Owned exposed control token was not rotated")
    current_pods = pods(h, CONTROL_NS)
    require(current_pods and not old_control_pods.intersection(pod["metadata"]["uid"] for pod in current_pods),
            "A startup-cached old control-token consumer survived Ready")
    expected = fingerprint(rotated)
    for pod in current_pods:
        require(pod["metadata"].get("annotations", {}).get(EPOCH) == reg["status"]["privacyEpoch"],
                "Control consumer lacks the new privacy epoch")
        cached = h.k("exec", "-n", CONTROL_NS, pod["metadata"]["name"], "-c", "credential-consumer",
                     "--", "cat", "/proof/cached.sha", timeout=25).split()[0]
        require(cached == expected, "Restarted control consumer did not cache the rotated token")
    h.passed("Owned router-services-admin token rotates and every cached old consumer is gone before Ready")
    h.cli("install", timeout=240)
    rollback_guard(h)
    h.state["legacy_registration_uid"] = reg["metadata"]["uid"]
    h.save()


def rollback_guard(h):
    # Execute the actual CLI guard against this API, without pretending a Kind
    # cluster has an Azure upgrade context or mocking an Azure command.
    script = """
const {pathToFileURL}=require('node:url');
const {execa}=await import(pathToFileURL(process.cwd()+'/cli/node_modules/execa/index.js'));
const {assertRollbackSafe}=await import(pathToFileURL(process.cwd()+'/cli/dist/lib/sre-authority.js'));
const execute=(file,args,options)=>execa(file,['--context','kind-kars-e2e',...args],options);
try { await assertRollbackSafe(execute); process.exitCode=9; }
catch(error) { if(!String(error.message).includes('Rollback across enrolled SRE authority is unsafe')) throw error; }
"""
    # CommonJS eval cannot use top-level await; the async wrapper keeps imports
    # on the actual compiled CLI module, not an alternate acceptance model.
    h.run(["node", "-e", f"(async()=>{{{script}}})().catch(()=>{{process.exitCode=1}})"], timeout=30)
    h.passed("Actual CLI rollback guard rejects regranting a legacy release while enrollment audit history exists")


def retire_and_uninstall(h):
    reg = h.get("karssreregistrations.kars.azure.com", "canonical")
    h.cli("authority", "retire", "--registration-uid", reg["metadata"]["uid"],
          "--resource-version", reg["metadata"]["resourceVersion"], user="registrar", timeout=240)
    retired = h.get("karssreregistrations.kars.azure.com", "canonical")
    require(retired["status"]["phase"] == "Retired" and retired["spec"]["enabled"] is False,
            "Retirement did not complete")
    for name in (PRIVATE, AGENT):
        require(h.get("secret", name, RUNTIME) is None, "Private credential survived retirement")
    for kind in ("clusterrolebindings", "rolebindings"):
        items = json.loads(h.k("get", kind, "-A", "-o", "json"))["items"]
        require(not any(item["metadata"].get("annotations", {}).get("kars.azure.com/sre-registration-uid") == reg["metadata"]["uid"]
                        for item in items), "Owned private grant survived retirement")
    h.cli("uninstall", timeout=180)
    h.poll("SRE source removal", lambda: h.get("karssandbox", "sre", SYSTEM) is None, seconds=120)
    h.poll("SRE namespace removal", lambda: h.get("namespace", RUNTIME) is None, seconds=120)
    require(h.get("namespace", SYSTEM)["metadata"]["uid"] == h.state["system_uid"], "Retirement changed core namespace identity")
    retained = h.get("karssreregistrations.kars.azure.com", "canonical")
    require(retained["metadata"]["uid"] == reg["metadata"]["uid"] and retained["status"]["phase"] == "Retired",
            "Retired registration audit record was removed or replaced")
    h.passed("Retire/uninstall revoke owned grants/credentials and remove the source namespace while retaining core UID and audit registration")


def fresh_reenrollment(h):
    policies_ready(h)
    retained = h.get("karssreregistrations.kars.azure.com", "canonical")
    require(retained and retained["status"]["phase"] == "Retired", "Fresh phase requires retained retired history")
    foreign = h.create({"apiVersion": "v1", "kind": "Namespace",
                        "metadata": {"name": RUNTIME, "labels": {"e2e-foreign-occupant": "true"}}})
    h.create({"apiVersion": "v1", "kind": "ConfigMap", "metadata": {"name": "foreign-sentinel", "namespace": RUNTIME},
              "data": {"value": "must-remain"}})
    cli_failure(h.cli("authority", "stage-source", expected=None),
                "Existing kars-sre namespace", "foreign occupant")
    require(h.get("namespace", RUNTIME)["metadata"]["uid"] == foreign["metadata"]["uid"], "Foreign namespace was replaced")
    require(h.get("configmap", "foreign-sentinel", RUNTIME)["data"]["value"] == "must-remain", "Foreign content was mutated")
    require(h.get("karssandbox", "sre", SYSTEM) is None and h.get("secret", PRIVATE, RUNTIME) is None,
            "Foreign occupancy acquired a source or private privilege")
    h.namespace_delete(RUNTIME, foreign["metadata"]["uid"])
    h.passed("Foreign runtime occupancy blocks fresh staging before source creation or private issuance")

    created = h.cli("authority", "stage-source")
    spec = enrollment_json(h.cli("authority", "preview", user="registrar"))
    assert_claim(h.get("karssandbox", "sre", SYSTEM), h.get("namespace", RUNTIME),
                 h.get("namespace", SYSTEM), h.state["system_uid"])
    require(f'--sandbox-uid {spec["sandbox"]["uid"]}' in created
            and f'--namespace-uid {spec["runtimeNamespace"]["uid"]}' in created,
            "Fresh staging did not return the actual CREATE/claim UIDs")
    require(spec["sandbox"]["uid"] != h.state["legacy_source_uid"] and
            spec["runtimeNamespace"]["uid"] != h.state["legacy_namespace_uid"], "Fresh source reused old authority identities")
    cli_failure(h.cli("authority", "stage-source", expected=None), "SRE source already exists", "source adoption race")
    cli_failure(h.cli("authority", "enroll", "--sandbox-uid", h.state["legacy_source_uid"],
                     "--namespace-uid", h.state["legacy_namespace_uid"], user="registrar", expected=None),
                "no longer matches", "stale source UID enrollment")
    retained = h.get("karssreregistrations.kars.azure.com", "canonical")
    reviewed = review_args(spec, retained)
    dry = printed_object(h.cli("authority", "enroll", *reviewed, "--dry-run", user="registrar"))
    require(dry["spec"] == spec, "Fresh enrollment differs from its reviewed identities")
    h.cli("authority", "enroll", *reviewed, user="registrar")
    require(h.get("karssreregistrations.kars.azure.com", "canonical")["spec"] == spec,
            "Fresh CAS re-enrollment retained unreviewed identities from the retired registration")
    h.cli("authority", "migrate", user="registrar", timeout=240)
    h.wait_ready()
    h.cli("install", timeout=240)
    # Recreate only our test Role/Bindings in the newly claimed namespace.
    role = h.get("role", "e2e-sre-admission-probe", TENANT)
    h.create({"apiVersion": role["apiVersion"], "kind": "Role",
              "metadata": {"name": role["metadata"]["name"], "namespace": RUNTIME}, "rules": role["rules"]})
    for account, namespace in (("tenant", TENANT), ("registrar", OPERATORS)):
        h.create({"apiVersion": "rbac.authorization.k8s.io/v1", "kind": "RoleBinding",
            "metadata": {"name": f"e2e-sre-{account}-runtime", "namespace": RUNTIME},
            "roleRef": {"apiGroup": "rbac.authorization.k8s.io", "kind": "Role", "name": role["metadata"]["name"]},
            "subjects": [{"kind": "ServiceAccount", "name": account, "namespace": namespace}]})
    h.passed("Fresh atomic source/claim and delegated CAS re-enrollment reach Ready without adopting stale UIDs")

File: controller/src/sre_authority/migration.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use super::{api_error, check_secret_denial};
use crate::{
    crd::KarsSandbox,
    sre_registration::{EPOCH, KarsSRERegistration, OWNER, RUNTIME_NAMESPACE},
};
use k8s_openapi::api::{
    apps::v1::Deployment,
    core::v1::{Namespace, Pod, Secret},
};
use kube::{
    Api, Client, ResourceExt,
    api::{DeleteParams, ListParams, Patch, PatchParams, Preconditions},
};
use serde_json::json;

const STOPPED: &str = "kars.azure.com/sre-migration-stopped";
pub(super) const WAITING_FOR_CONSUMERS: &str =
    "Waiting for legacy SRE consumers to terminate; no private credentials issued";
const WAITING_FOR_ROTATION: &str =
    "Waiting for owned control credential consumers to restart on the new privacy epoch";
const WAITING_FOR_ROLLOUT: &str =
    "Owned control credential consumer has not completed its privacy-epoch rollout";
const WAITING_FOR_CONSUMER_REMOVAL: &str =
    "Waiting for the owned SRE consumer Deployment to be removed";

pub(super) fn is_waiting(detail: &str) -> bool {
    matches!(
        detail,
        WAITING_FOR_CONSUMERS
            | WAITING_FOR_ROTATION
            | WAITING_FOR_ROLLOUT
            | WAITING_FOR_CONSUMER_REMOVAL
    )
}

pub(super) async fn stop_registered_consumer_for_retirement(
    client: &Client,
    reg: &KarsSRERegistration,
) -> Result<(), String> {
    let namespaces: Api<Namespace> = Api::all(client.clone());
    let Some(namespace) = namespaces
        .get_opt(RUNTIME_NAMESPACE)
        .await
        .map_err(|e| api_error("Read retiring SRE namespace", e))?
    else {
        return Ok(());
    };
    if namespace.metadata.uid.as_deref() != Some(reg.spec.runtime_namespace.uid.as_str()) {
        return Ok(());
    }
    let registration_uid = reg
        .metadata
        .uid
        .as_deref()
        .filter(|uid| !uid.is_empty())
        .ok_or("SRE retirement requires the actual registration UID")?;
    let api: Api<Deployment> = Api::namespaced(client.clone(), RUNTIME_NAMESPACE);
    let Some(deployment) = api
        .get_opt("sre")
        .await
        .map_err(|e| api_error("Read retiring SRE consumer", e))?
    else {
        return Ok(());
    };
    let ours = deployment
        .spec
        .as_ref()
        .and_then(|s| s.template.metadata.as_ref())
        .and_then(|m| m.annotations.as_ref())
        .and_then(|a| a.get(OWNER))
        .map(String::as_str)
        == Some(registration_uid);
    if !ours
        && reg
            .status
            .as_ref()
            .and_then(|s| s.router_service_account_uid.as_ref())
            .is_none()
    {
        return Ok(());
    }
    stop_legacy_consumer(client, reg).await?;
    if !ours {
        return Ok(());
    }
    // Namespace-controller intentionally lacks registrar power. Remove only
    // our stopped Deployment here instead of leaving protected content to GC.
    let Some(current) = api
        .get_opt("sre")
        .await
        .map_err(|error| api_error("Read stopped SRE consumer", error))?
    else {
        return Ok(());
    };
    let uid = deployment
        .metadata
        .uid
        .as_deref()
        .filter(|uid| !uid.is_empty())
        .ok_or("Retiring SRE consumer UID is missing")?;
    let still_owned = current
        .spec
        .as_ref()
        .and_then(|spec| spec.template.metadata.as_ref())
        .and_then(|metadata| metadata.annotations.as_ref())
        .and_then(|annotations| annotations.get(OWNER))
        .map(String::as_str)
        == Some(registration_uid);
    if current.metadata.uid.as_deref() != Some(uid)
        || !still_owned
        || current.spec.as_ref().and_then(|spec| spec.replicas) != Some(0)
    {
        return Err("Retiring SRE consumer changed after quiescence; preserved".into());
    }
    if current.metadata.deletion_timestamp.is_some() {
        return Err(WAITING_FOR_CONSUMER_REMOVAL.into());
    }
    let version = current
        .metadata
        .resource_version
        .filter(|version| !version.is_empty())
        .ok_or("Retiring SRE consumer resourceVersion is missing")?;
    match api
        .delete(
            "sre",
            &DeleteParams {
                preconditions: Some(Preconditions {
                    uid: Some(uid.into()),
                    resource_version: Some(version),
                }),
                ..Default::default()
            },
        )
        .await
    {
        Ok(_) => {}
        Err(kube::Error::Api(error)) if error.code == 404 => {}
        Err(error) => return Err(api_error("Remove stopped owned SRE consumer", error)),
    }
    if api
        .get_opt("sre")
        .await
        .map_err(|error| api_error("Verify owned SRE consumer removal", error))?
        .is_some()
    {
        return Err(WAITING_FOR_CONSUMER_REMOVAL.into());
    }
    Ok(())
}

fn current_boundary(deployment: &Deployment, reg: &KarsSRERegistration) -> bool {
    deployment.spec.as_ref().is_some_and(|spec| {
        spec.template
            .metadata
            .as_ref()
            .and_then(|m| m.annotations.as_ref())
            .is_some_and(|a| {
                a.get(OWNER) == reg.metadata.uid.as_ref() && a.get(EPOCH) == Some(&reg.epoch())
            })
    })
}

pub(super) async fn validate_consumer(
    client: &Client,
    reg: &KarsSRERegistration,
) -> Result<(), String> {
    let api: Api<Deployment> = Api::namespaced(client.clone(), RUNTIME_NAMESPACE);
    let Some(deployment) = api
        .get_opt("sre")
        .await
        .map_err(|e| api_error("Read SRE consumer", e))?
    else {
        return Ok(());
    };
    if current_boundary(&deployment, reg) {
        return Ok(());
    }
    if deployment
        .spec
        .as_ref()
        .and_then(|s| s.template.metadata.as_ref())
        .and_then(|m| m.annotations.as_ref())
        .and_then(|a| a.get(OWNER))
        == reg.metadata.uid.as_ref()
    {
        return Ok(());
    }
    let reviewed = reg
        .spec
        .legacy_consumer
        .as_ref()
        .ok_or("Existing SRE consumer requires an explicit UID/resourceVersion review")?;
    let was_stopped = deployment
        .metadata
        .annotations
        .as_ref()
        .and_then(|a| a.get(STOPPED))
        == reg.metadata.uid.as_ref();
    if deployment.metadata.uid.as_deref() != Some(reviewed.uid.as_str())
        || deployment.metadata.deletion_timestamp.is_some()
        || (!was_stopped
            && deployment.metadata.resource_version.as_deref()
                != Some(reviewed.resource_version.as_str()))
    {
        return Err("Reviewed SRE consumer was changed or replaced".into());
    }
    Ok(())
}

pub(super) async fn stop_legacy_consumer(
    client: &Client,
    reg: &KarsSRERegistration,
) -> Result<(), String> {
    let api: Api<Deployment> = Api::namespaced(client.clone(), RUNTIME_NAMESPACE);
    if let Some(deployment) = api
        .get_opt("sre")
        .await
        .map_err(|e| api_error("Read SRE consumer for retirement", e))?
    {
        if reg.spec.enabled && current_boundary(&deployment, reg) {
            return Ok(());
        }
        validate_consumer(client, reg).await?;
        if deployment.spec.as_ref().and_then(|spec| spec.replicas) != Some(0) {
            api.patch("sre",&PatchParams::default(),&Patch::Merge(json!({
                "metadata":{"uid":deployment.metadata.uid,"resourceVersion":deployment.metadata.resource_version,
                    "annotations":{STOPPED:reg.metadata.uid}},
                "spec":{"replicas":0},
            }))).await.map_err(|e|api_error("Stop reviewed SRE consumer",e))?;
        }
    }
    let pods: Api<Pod> = Api::namespaced(client.clone(), RUNTIME_NAMESPACE);
    let pods = pods
        .list(&ListParams::default())
        .await
        .map_err(|e| api_error("Wait for old SRE consumers", e))?;
    if pods.iter().any(|pod| {
        pod.spec
            .as_ref()
            .and_then(|spec| spec.service_account_name.as_deref())
            == Some("sandbox")
    }) {
        return Err(WAITING_FOR_CONSUMERS.into());
    }
    Ok(())
}

fn controller_managed(deployment: &Deployment, sandbox: &str) -> bool {
    deployment
        .metadata
        .labels
        .as_ref()
        .and_then(|labels| labels.get("kars.azure.com/sandbox"))
        .map(String::as_str)
        == Some(sandbox)
        && deployment
            .metadata
            .managed_fields
            .as_ref()
            .is_some_and(|fields| {
                fields.iter().any(|field| {
                    field.manager.as_deref() == Some(crate::field_managers::CLAWSANDBOX)
                        && field
                            .fields_v1
                            .as_ref()
                            .is_some_and(|fields| fields.0.get("f:spec").is_some())
                })
            })
}

/// Existing #550 control credentials may have been exposed by the old SRE
/// identity. Rotate only proven controller-owned instances and restart their
/// owned consumer; the old router caches the token at startup.
pub(super) async fn rotate_owned_control_credentials(
    client: &Client,
    reg: &KarsSRERegistration,
) -> Result<(), String> {
    let all: Api<Secret> = Api::all(client.clone());
    let list = all
        .list(&ListParams::default().fields("metadata.name=router-services-admin"))
        .await
        .map_err(|e| api_error("Inventory owned control credentials", e))?;
    for secret in list {
        if secret
            .metadata
            .labels
            .as_ref()
            .and_then(|labels| labels.get("app.kubernetes.io/managed-by"))
            .map(String::as_str)
            != Some("kars-controller")
        {
            continue;
        }
        if secret.metadata.uid.as_deref().is_none_or(str::is_empty)
            || secret
                .metadata
                .resource_version
                .as_deref()
                .is_none_or(str::is_empty)
            || secret.metadata.deletion_timestamp.is_some()
            || secret
                .metadata
                .owner_references
                .as_ref()
                .is_some_and(|owners| !owners.is_empty())
        {
            return Err("Owned control credential identity is incomplete or terminating".into());
        }
        let Some(namespace_name) = secret.namespace() else {
            continue;
        };
        let annotations = secret.metadata.annotations.as_ref();
        let Some(source_uid) = annotations.and_then(|a| a.get("kars.azure.com/sandbox-uid")) else {
            continue;
        };
        let Some(namespace_uid) = annotations.and_then(|a| a.get("kars.azure.com/namespace-uid"))
        else {
            continue;
        };
        let namespaces: Api<Namespace> = Api::all(client.clone());
        let namespace = namespaces
            .get(&namespace_name)
            .await
            .map_err(|e| api_error("Read control credential namespace", e))?;
        if namespace.metadata.uid.as_deref() != Some(namespace_uid)
            || namespace.metadata.deletion_timestamp.is_some()
        {
            return Err("Control credential namespace changed".into());
        }
        let ns_annotations = namespace
            .metadata
            .annotations
            .as_ref()
            .ok_or("Control namespace claim is absent")?;
        let workspace = ns_annotations
            .get("kars.azure.com/sandbox-namespace")
            .ok_or("Control namespace source is absent")?;
        let name = ns_annotations
            .get("kars.azure.com/sandbox-name")
            .ok_or("Control namespace owner is absent")?;
        let sandboxes: Api<KarsSandbox> = Api::namespaced(client.clone(), workspace);
        let sandbox = sandboxes
            .get(name)
            .await
            .map_err(|e| api_error("Read control credential owner", e))?;
        if sandbox.metadata.uid.as_deref() != Some(source_uid)
            || sandbox.metadata.deletion_timestamp.is_some()
            || !crate::reconciler::namespace_ownership::claimed(&namespace, &sandbox)
                .map_err(|_| "Control namespace ownership is invalid")?
        {
            return Err("Control credential ownership changed; no rotation was authorized".into());
        }
        check_secret_denial(client, &namespace_name).await?;
        let deployments: Api<Deployment> = Api::namespaced(client.clone(), &namespace_name);
        let Some(deployment) = deployments
            .get_opt(name)
            .await
            .map_err(|e| api_error("Read control credential consumer", e))?
        else {
            return Err("Owned control credential has no verified consumer Deployment".into());
        };
        if !controller_managed(&deployment, name) {
            return Err("Control credential consumer is not controller-owned".into());
        }
        let epoch = reg.epoch();
        if annotations.and_then(|a| a.get(EPOCH)) != Some(&epoch) {
            let secrets: Api<Secret> = Api::namespaced(client.clone(), &namespace_name);
            secrets.patch("router-services-admin",&PatchParams::default(),&Patch::Merge(json!({
                "metadata":{"uid":secret.metadata.uid,"resourceVersion":secret.metadata.resource_version,
                    "annotations":{EPOCH:epoch}},
                "stringData":{"control-token":crate::providers::signing::generate_service_token()},
            }))).await.map_err(|e|api_error("Rotate owned control credential",e))?;
        }
        if deployment
            .spec
            .as_ref()
            .and_then(|spec| spec.template.metadata.as_ref())
            .and_then(|meta| meta.annotations.as_ref())
            .and_then(|a| a.get(EPOCH))
            != Some(&epoch)
        {
            deployments.patch(name,&PatchParams::default(),&Patch::Merge(json!({
                "metadata":{"uid":deployment.metadata.uid,"resourceVersion":deployment.metadata.resource_version},
                "spec":{"template":{"metadata":{"annotations":{EPOCH:epoch}}}},
            }))).await.map_err(|e|api_error("Restart owned control credential consumer",e))?;
            return Err(WAITING_FOR_ROTATION.into());
        }
        let desired = deployment
            .spec
            .as_ref()
            .and_then(|spec| spec.replicas)
            .unwrap_or(1);
        let ready = deployment.status.as_ref().is_some_and(|status| {
            status.observed_generation == deployment.metadata.generation
                && status.updated_replicas.unwrap_or(0) == desired
                && status.available_replicas.unwrap_or(0) == desired
        });
        if !ready {
            return Err(WAITING_FOR_ROLLOUT.into());
        }
        let selector = deployment
            .spec
            .as_ref()
            .and_then(|spec| spec.selector.match_labels.as_ref())
            .filter(|labels| !labels.is_empty())
            .ok_or("Owned control credential consumer has no bounded Pod selector")?
            .iter()
            .map(|(key, value)| format!("{key}={value}"))
            .collect::<Vec<_>>()
            .join(",");
        let pods = Api::<Pod>::namespaced(client.clone(), &namespace_name)
            .list(&ListParams::default().labels(&selector))
            .await
            .map_err(|e| api_error("Verify old control credential consumers have terminated", e))?;
        if pods.iter().any(|pod| {
            pod.metadata
                .annotations
                .as_ref()
                .and_then(|annotations| annotations.get(EPOCH))
                != Some(&epoch)
        }) {
            // Rollout availability alone can exclude a still-terminating old
            // router which continues accepting its startup-cached control token.
            return Err(WAITING_FOR_ROLLOUT.into());
        }
    }
    Ok(())
}
